pkg/apis/pipeline/v1beta1/param_types.go

```diff
@@ -38,40 +38,64 @@ type ParamSpec struct {
 	// Name declares the name by which a parameter is referenced.
 	Name string `json:"name"`
 	// Type is the user-specified type of the parameter. The possible types
-	// are currently "string" and "array", and "string" is the default.
+	// are currently "string", "array" and "object", and "string" is the default.
 	// +optional
 	Type ParamType `json:"type,omitempty"`
 	// Description is a user-facing description of the parameter that may be
 	// used to populate a UI.
 	// +optional
 	Description string `json:"description,omitempty"`
+	// Properties is the JSON Schema properties to support key-value pairs parameter.
+	// +optional
+	Properties map[string]PropertySpec `json:"properties,omitempty"`
 	// Default is the value a parameter takes if no input value is supplied. If
 	// default is set, a Task may be executed without a supplied value for the
 	// parameter.
 	// +optional
 	Default *ArrayOrString `json:"default,omitempty"`
 }
 
+// PropertySpec defines the struct for object keys
+type PropertySpec struct {
+	Type ParamType `json:"type,omitempty"`
+}
+
 // SetDefaults set the default type
 func (pp *ParamSpec) SetDefaults(ctx context.Context) {
-	if pp != nil && pp.Type == "" {
-		if pp.Default != nil {
-			// propagate the parsed ArrayOrString's type to the parent ParamSpec's type
-			if pp.Default.Type != "" {
-				// propagate the default type if specified
-				pp.Type = pp.Default.Type
+	if pp == nil || pp.Type != "" {
+		return
+	}
+
+	// if the properties section is not empty, set object as the parameter type
+	if pp.Properties != nil {
+		pp.Type = ParamTypeObject
+		for _, property := range pp.Properties {
+			property.Type = ParamTypeString
+		}
+		return
+	}
+
+	if pp.Default != nil {
+		// propagate the parsed ArrayOrString's type to the parent ParamSpec's type
+		if pp.Default.Type != "" {
+			// propagate the default type if specified
+			pp.Type = pp.Default.Type
+		} else {
+			// determine the type based on the array or string values when default value is specified but not the type
+			if pp.Default.ArrayVal != nil {
+				pp.Type = ParamTypeArray
 			} else {
-				// determine the type based on the array or string values when default value is specified but not the type
-				if pp.Default.ArrayVal != nil {
-					pp.Type = ParamTypeArray
+				// not using else-if to avoid compaint from gocritic
```

Comment:
It will suggest to use `switch` when `if-else` chains is detected, which will fail the build tests. And apparently this code chuck cannot be done with switch because we are checking different variables in the struct.

Reply:
@wlynch that works! Thanks!